src/main/java/frc/robot/commands/DriveStraight.java

```diff
@@ -0,0 +1,78 @@
+package frc.robot.commands;
+
+import edu.wpi.first.wpilibj.PIDController;
+import edu.wpi.first.wpilibj.PIDSource;
+import edu.wpi.first.wpilibj.Timer;
+import edu.wpi.first.wpilibj.command.Command;
+import frc.robot.PidSettings;
+import frc.robot.Robot;
+import frc.robot.RobotComponents;
+import frc.robot.RobotConstants;
+import frc.robot.pidsources.DrivetrainPidSource;
+
+/** Drive straight with encoders and gyro. */
+public class DriveStraight extends Command {
+  private PidSettings pidSettingsX, pidSettingsY;
+  private PIDController pidControllerX, pidControllerY;
+  private double timeOnTarget, setpoint, outputX, outputY;
+  private PIDSource drivePidSource;
+
+  public DriveStraight(double setpoint, PidSettings pidSettingsX, PidSettings PidSettingsY) {
+    requires(Robot.drivetrain);
+    this.setpoint = setpoint;
+    this.pidSettingsX = pidSettingsX;
+    this.pidSettingsY = pidSettingsY;
+  }
+
+  public DriveStraight(double setpoint) {
+    this(setpoint, RobotConstants.RobotPIDSettings.DRIVETRAIN_TURN_PID_SETTINGS,
+        RobotConstants.RobotPIDSettings.DRIVETRAIN_ENCODERS_PID_SETTINGS);
+  }
+
+  @Override
+  protected void initialize() {
+    drivePidSource = new DrivetrainPidSource();
+    this.pidControllerY = new PIDController(pidSettingsY.getKP(), pidSettingsY.getKI(), pidSettingsY.getKD(),
+        drivePidSource, output -> outputY = output);
+    pidControllerY.setOutputRange(-1, 1);
+    pidControllerY.setAbsoluteTolerance(pidSettingsY.getTolerance());
+    pidControllerY.setSetpoint(this.setpoint);
+
+    
+    this.pidControllerX = new PIDController(pidSettingsX.getKP(), pidControllerX.getI(), pidControllerX.getD(),
+        RobotComponents.Drivetrain.GYRO, output -> outputX = output);
+    pidControllerY.setOutputRange(-1, 1);
+    pidControllerY.setAbsoluteTolerance(pidSettingsY.getTolerance());
+    pidControllerY.setSetpoint(0);
```

Comment:
you got confused between x and y.
setpoint should be `Robot.drivetrain.getAngle()` and not 0.